File: src/MagnusAddon/note/vocabulary/vocabnote_usercompoundparts.py
from __future__ import annotations

from typing import TYPE_CHECKING, override

from ex_autoslot import ProfilableAutoSlots
from language_services.jamdict_ex.dict_lookup import DictLookup
from language_services.janome_ex.word_extraction.word_exclusion import WordExclusion
from note.note_constants import NoteFields
from note.notefields.comma_separated_strings_list_field import MutableCommaSeparatedStringsListField
from note.sentences.sentence_configuration import SentenceConfiguration

if TYPE_CHECKING:
    from note.collection.jp_collection import JPCollection
    from note.vocabulary.vocabnote import VocabNote
    from sysutils.weak_ref import WeakRef

class VocabNoteUserCompoundParts(ProfilableAutoSlots):
    def __init__(self, vocab: WeakRef[VocabNote]) -> None:
        self.__vocab = vocab
        self._field: MutableCommaSeparatedStringsListField = MutableCommaSeparatedStringsListField(vocab, NoteFields.Vocab.user_compounds)

    @property
    def _vocab(self) -> VocabNote: return self.__vocab()

    @property
    def _collection(self) -> JPCollection: return self._vocab.collection

    def primary(self) -> list[str]: return [part for part in self._field.get() if not part.startswith("[")]
    def all(self) -> list[str]: return [self._strip_brackets(part) for part in self._field.get()]
    def set(self, value: list[str]) -> None: self._field.set(value)

    def auto_generate(self) -> None:
        from language_services.janome_ex.word_extraction.text_analysis import TextAnalysis
        from note.vocabulary.vocabnote import VocabNote
        analysis = TextAnalysis(self._vocab.get_question(), SentenceConfiguration.from_incorrect_matches([WordExclusion.global_(form) for form in self._vocab.forms.all_set()]))
        compound_parts = [a.form for a in analysis.display_word_variants if a.form not in self._vocab.forms.all_set()]
        if not len(compound_parts) > 1:  # time to brute force it
            word = self._vocab.get_question()
            all_substrings = [word[i:j] for i in range(len(word)) for j in range(i + 1, len(word) + 1) if word[i:j] != word]
            all_word_substrings = [w for w in all_substrings if DictLookup.is_dictionary_or_collection_word(w)]
            compound_parts = [segment for segment in all_word_substrings if not any(parent for parent in all_word_substrings if segment in parent and parent != segment)]

        segments_missing_vocab = [segment for segment in compound_parts if not self._collection.vocab.is_word(segment)]
        for missing in segments_missing_vocab:
            created = VocabNote.factory.create_with_dictionary(missing)
            created.suspend_all_cards()

        self.set(compound_parts)

    @staticmethod
    def _strip_brackets(part: str) -> str:
        return part.replace("[", "").replace("]", "")

    @override
    def __repr__(self) -> str: return self._field.__repr__()

File: src/src_dotnet/JAStudio.Core/Note/Vocabulary/VocabNoteUserCompoundParts.cs
using System.Collections.Generic;
using System.Linq;
using JAStudio.Core.LanguageServices.JanomeEx;
using JAStudio.Core.LanguageServices.JanomeEx.WordExtraction;
using JAStudio.Core.Note.Collection;
using JAStudio.Core.Note.NoteFields;
using JAStudio.Core.Note.Sentences;

namespace JAStudio.Core.Note.Vocabulary;

public class VocabNoteUserCompoundParts
{
    private readonly VocabNote _vocab;
    private readonly MutableCommaSeparatedStringsListField _field;

    public VocabNoteUserCompoundParts(VocabNote vocab)
    {
        _vocab = vocab;
        _field = new MutableCommaSeparatedStringsListField(vocab, NoteFieldsConstants.Vocab.UserCompounds);
    }

    private VocabNote Vocab => _vocab;
    private JPCollection Collection => Vocab.Collection;

    public List<string> Primary()
    {
        return _field.Get()
            .Where(part => !part.StartsWith("["))
            .ToList();
    }

    public List<string> All()
    {
        return _field.Get()
            .Select(StripBrackets)
            .ToList();
    }

    /// Returns all compound parts with brackets preserved (e.g. "[word]" stays as-is).
    public List<string> AllRaw() => _field.Get();

    public void Set(List<string> value)
    {
        _field.Set(value);
    }

    public HashSet<VocabNote> AllNotes()
    {
        return All()
            .SelectMany(part => Vocab.Services.Collection.Vocab.WithQuestion(part))
            .ToHashSet();
    }

    public List<VocabNote> PrimaryPartsNotes()
    {
        return Primary()
            .SelectMany(part => Vocab.Services.Collection.Vocab.WithFormPreferDisambiguationNameOrExactMatch(part))
            .ToList();
    }

    public void AutoGenerate()
    {
        var exclusions = Vocab.Forms.AllSet()
            .Select(form => WordExclusion.Global(form))
            .ToList();
        var config = SentenceConfiguration.FromIncorrectMatches(exclusions);
        var analysisServices = new AnalysisServices(Vocab.Services.Collection.Vocab, Vocab.Services.DictLookup, Vocab.Services.Settings);
        var analysis = new TextAnalysis(analysisServices, Vocab.GetQuestion(), config);
        var compoundParts = analysis.DisplayWordVariants
            .Where(a => !Vocab.Forms.AllSet().Contains(a.Form))
            .Select(a => a.Form)
            .ToList();

        if (compoundParts.Count <= 1)  // time to brute force it
        {
            var word = Vocab.GetQuestion();
            var allSubstrings = new List<string>();
            for (int i = 0; i < word.Length; i++)
            {
                for (int j = i + 1; j <= word.Length; j++)
                {
                    var substring = word.Substring(i, j - i);
                    if (substring != word)
                    {
                        allSubstrings.Add(substring);
                    }
                }
            }

            var allWordSubstrings = allSubstrings
                .Where(w => Vocab.Services.DictLookup.IsDictionaryOrCollectionWord(w))
                .ToList();

            compoundParts = allWordSubstrings
                .Where(segment => !allWordSubstrings.Any(parent => parent.Contains(segment) && parent != segment))
                .ToList();
        }

        var segmentsMissingVocab = compoundParts
            .Where(segment => !Collection.Vocab.IsWord(segment))
            .ToList();

        foreach (var missing in segmentsMissingVocab)
        {
            Vocab.Services.VocabNoteFactory.CreateWithDictionary(missing);
        }

        Set(compoundParts);
    }

    private static string StripBrackets(string part)
    {
        return part.Replace("[", string.Empty).Replace("]", string.Empty);
    }

    public override string ToString()
    {
        return _field.ToString() ?? string.Empty;
    }
}
